frame/democracy/src/lib.rs

```diff
@@ -0,0 +1,2084 @@
+		/// proposed again.
+		///
+		/// If called on a queued public or external proposal, then this will result in it being
+		/// removed. If the `ref_index` supplied is an active referendum with the proposal hash,
+		/// then it will be cancelled.
+		///
+		/// The dispatch origin of this call must be `BlacklistOrigin`.
+		///
+		/// - `proposal_hash`: The proposal hash to blacklist permanently.
+		/// - `ref_index`: An ongoing referendum whose hash is `proposal_hash`, which will be
+		/// cancelled.
+		///
+		/// Weight: `O(p)` (though as this is an high-privilege dispatch, we assume it has a
+		///   reasonable value).
+		#[pallet::weight((T::WeightInfo::blacklist(T::MaxProposals::get()), DispatchClass::Operational))]
+		#[transactional]
+		pub fn blacklist(
+			origin: OriginFor<T>,
+			proposal_hash: T::Hash,
+			asset_id: T::AssetId,
+			maybe_ref_index: Option<ReferendumIndex>,
+		) -> DispatchResult {
+			T::BlacklistOrigin::ensure_origin(origin)?;
+
+			let id = ProposalId { hash: proposal_hash, asset_id };
+
+			// Insert the proposal into the blacklist.
+			let permanent = (T::BlockNumber::max_value(), Vec::<T::AccountId>::new());
+			Blacklist::<T>::insert(&id, permanent);
+
+			// Remove the queued proposal, if it's there.
+			PublicProps::<T>::try_mutate(|props| -> DispatchResult {
+				if let Some(index) = props.iter().position(|p| p.1 == id) {
+					let (prop_index, ..) = props.remove(index);
+					if let Some((whos, amount)) = DepositOf::<T>::take(prop_index) {
+						for who in whos.into_iter() {
+							T::NativeCurrency::transfer_held(
+								&who,
+								&T::TreasuryAccount::get(),
+								amount,
+								true,
+								false,
+							)?;
+						}
+					}
+				}
+				Ok(())
+			})?;
+
+			// Remove the external queued referendum, if it's there.
+			if matches!(NextExternal::<T>::get(), Some((h, ..)) if h == id) {
+				NextExternal::<T>::kill();
+			}
+
+			// Remove the referendum, if it's there.
+			if let Some(ref_index) = maybe_ref_index {
+				if let Ok(status) = Self::referendum_status(ref_index) {
+					if status.proposal_id == id {
+						Self::internal_cancel_referendum(ref_index);
+					}
+				}
+			}
+
+			Self::deposit_event(Event::<T>::Blacklisted(proposal_hash));
+			Ok(())
+		}
+
+		/// Remove a proposal.
+		///
+		/// The dispatch origin of this call must be `CancelProposalOrigin`.
+		///
+		/// - `prop_index`: The index of the proposal to cancel.
+		///
+		/// Weight: `O(p)` where `p = PublicProps::<T>::decode_len()`
+		#[pallet::weight(T::WeightInfo::cancel_proposal(T::MaxProposals::get()))]
+		pub fn cancel_proposal(
+			origin: OriginFor<T>,
+			#[pallet::compact] prop_index: PropIndex,
+		) -> DispatchResult {
+			T::CancelProposalOrigin::ensure_origin(origin)?;
+
+			PublicProps::<T>::mutate(|props| props.retain(|p| p.0 != prop_index));
+			if let Some((whos, amount)) = DepositOf::<T>::take(prop_index) {
+				for who in whos.into_iter() {
+					T::NativeCurrency::transfer_held(
```

Comment:
it seems like it is possible to impl same using previos abstration? so code will be same?

Reply:
Yes, but the original code has it done separately so left it as-is.

Reply:
ok, just was thinkin cause we are not going too far from original, keep it as close as possible to ease merges.